src/server/pbsd_main.c

```diff
@@ -1677,6 +1677,15 @@ main(int argc, char **argv)
 
 	pbs_python_ext_start_interpreter(&svr_interp_data);
 
+	if (!svr_interp_data.pbs_python_types_loaded ||
+			!svr_interp_data.interp_started ||
+			!svr_interp_data.data_initialized) {
+		log_err(-1, msg_daemonname, "Failed to start python interpreter!");
+		stop_db();
```

Comment:
Thank you for your contribution!

It might make more sense to initialize the Python interpreter before we attempt to connect to the database. We can detect the problem and exit sooner, avoiding unnecessary work. What do you think?

When referring to Python (as a proper name, not a command) please capitalize it.

Reply:
@mike0042 : I believe we have to start the database first before starting the Python  interpreter, for the interpreter would need PBS attributes and resources info from database, in order to instantiate server, queue, node, vnode, job objects...

Reply:
Also, it looks like pbs_python_ext_start_interpreter call and supporting code would go around line 1251 but it looks like python is started and stopped somewhere below that with pbs_python_ext_quick_start_interpreter(1399) and pbs_python_ext_quick_shutdown_interpreter(1420).  

Updated python -> Python.

Reply:
@ericpershey How about we move this check inside `pbs_python_ext_start_interpreter` itself and change the signature of `pbs_python_ext_start_interpreter` to `int` from `void` and return failure (-1 or 1). Because `pbs_python_ext_start_interpreter` is called from few other places also in PBS code.

```
pbspro/src/tools/pbs_python.c:2909:		pbs_python_ext_start_interpreter(&svr_interp_data);
pbspro/src/server/node_func.c:1440:			pbs_python_ext_start_interpreter(&svr_interp_data);
pbspro/src/server/pbsd_main.c:1678:	pbs_python_ext_start_interpreter(&svr_interp_data);
pbspro/src/server/setup_resc.c:85:	pbs_python_ext_start_interpreter(&svr_interp_data);
pbspro/src/server/node_manager.c:3460:		pbs_python_ext_start_interpreter(&svr_interp_data);
pbspro/src/server/node_manager.c:4171:		pbs_python_ext_start_interpreter(&svr_interp_data);
pbspro/src/lib/Libpython/pbs_python_svr_internal.c:5246:		pbs_python_ext_start_interpreter(&svr_interp_data);
``` 

@mike0042 and @bayucan What do you guys say about moving failure check inside `pbs_python_ext_start_interpreter`?

Reply:
@ericpershey I don't think you need to return `(!svr_interp_data.pbs_python_types_loaded || !svr_interp_data.interp_started || !svr_interp_data.data_initialized)`, you can simply return 1 or 0.

Reply:
Either way is fine, IMHO. Zero for success, non-zero for failure.

Reply:
I'm fine with putting in the failure check inside pbs_python_ext_start_interpreter() and then going through each of the cases where the function is called and deal with having it or exit or just ignored.

Reply:
I have pushed changes.  There are a few places where we have open questions.  Please see the FIXME comments.

Reply:
All of the functions that call restart_python_interpreter() return void themselves, so aside from logging a message nothing else happens when a restart fails. For now, I think we should return whatever value pbs_python_ext_start_interpreter() returned from restart_python_interpreter() rather than void. At least that way the callers can choose to do something in the event of a failure in the future.

Reply:
So basically change the signature of restart_python_interpreter to return an int that is captured from pbs_python_ext_start_interpreter?

Reply:
Exactly.